A59-audit-logging.md

```diff
@@ -0,0 +1,281 @@
+          "description": "The typed config for the audit_log."
+            "This needs to be a json object mapped from google.protobuf.Struct"
+            "proto message.",
+          "type": "object",
+        }
+      },
+    }
+  },
+  "properties": {
+    ...
+    "audit_condition": {
+      "description": "Condition for the audit logging to be happen."
+        "Same as NONE if ommited.",
+      "type": "string",
+      "enum": ["NONE", "ON_DENY", "ON_ALLOW", "ON_DENY_AND_ALLOW"]
+    },
+    "audit_log": {
+      "$ref": "#/definitions/audit_log"
+    }
+  },
+  "required": ["name", "allow_rules"]
+}
+```
+
+For readers that are more familiar with proto messages, the change translates
+to:
+
+```
+message AuthorizationPolicy {
+  …
+
+  enum AuditCondition {
+    NONE = 0;
+    ON_DENY = 1;
+    ON_ALLOW = 2;
+    ON_DENY_AND_ALLOW = 3;
+  }
+
+  AuditCondition audit_condition = 1;
+
+  message AuditLog {
+    string name = 1;
+    google.protobuf.Struct config = 2;  
+  }
+  
+  AuditLog audit_log = 2;
+}
+```
+
+Note that the definition above is only for illustration purposes and does not
+actually exist anywhere as a `.proto` file. Nor does gRPC process the policy
+as protobuf by any means.
+
+Since the authorization policy is backed by two RBAC filters, it's necessary
+to clarify how this top-level audit condition is converted to two conditions
+in those RBACs.
+
+First of all, note that the RBAC with `DENY` is placed before the RBAC with
+`ALLOW`. We assume users will want to audit one particular RPC once if the
+condition is met.
+
+If users want to audit on deny, then both of the RBACs will have `ON_DENY` as
+the audit condition. The `ALLOW` RBAC will not be evaluated so audit logging
+will never happen twice. 
+
+If users want to audit on allow, then the `DENY` RBAC will have no audit enabled
+and the `ALLOW` RBAC with `ON_ALLOW`. So audit logging will at most happen once
+when the RPC passes through both RBACs.
+
+If users want to audit on both cases, then the `DENY` RBAC needs `ON_DENY` and
+the `ALLOW` RBAC needs `ON_DENY_AND_ALLOW`.
+
+Following is the table summarzing the combinations.
+
+|Authorization Policy  |DENY RBAC          |ALLOW RBAC           |
+|----------------------|-------------------|---------------------|
+|NONE                  |NONE               |NONE                 |
+|ON_DENY               |ON_DENY            |ON_DENY              |
+|ON_ALLOW              |NONE               |ON_ALLOW             |
+|ON_DENY_AND_ALLOW     |ON_DENY            |ON_DENY_AND_ALLOW    |
+
+Again this is a subset of what users can technically do with xDS RBACs in the way
+described in the earlier section. But we think this represents the most common
+use cases.
+
+### Language Specific APIs
+
+_This section is to be done once the language-agnostic APIs are more finalized._
+
+_What is certain for now is we need to support third-party audit logger
+implementations registered by users._
+
+## Rationale
```

Comment:
I really like how [https://github.com/grpc/proposal/blob/4187b46efa08eea782685efb91509ad77cc5e096/A43-grpc-authorization-api.md#rationale](same section in authorization) describes a high level flow from a user perspective. Do you think a similar section will add value here?

Reply:
Good idea. Working on it.

Reply:
Added some paragraphs. PTAL.